packages/whenable/src/whenable.js

```diff
@@ -0,0 +1,111 @@
+// @ts-check
+import { makePromiseKit } from '@endo/promise-kit';
+import { M } from '@endo/patterns';
+import { makeTagged } from '@endo/pass-style';
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ */
+export const prepareWhenableKits = zone => {
+  /** WeakMap<object, any> */
+  const whenableV0ToEphemeral = new WeakMap();
+
+  /**
+   * Get the current incarnation's promise kit associated with a whenableV0.
+   *
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @returns {import('@endo/promise-kit').PromiseKit<any>}
+   */
+  const findCurrentKit = whenableV0 => {
+    let pk = whenableV0ToEphemeral.get(whenableV0);
+    if (pk) {
+      return pk;
+    }
```

Comment:
We use this pattern so much that I'm a little shocked to not see conventional use of a helper like e.g. `const { cache: whenableV0ToEphemeral, provide: findCurrentKit } = makeCacheMapKit(_whenableV0 => makeQuietPromiseKit())` generalizing `weakMemoizeUnary`/`provideWeak`/`makeEphemeraProvider`/etc., but I guess it's easy to write modulo naming the function and [especially] the map (_which is usually ${keyDescription}To${ValDescription} as here, sometimes ${valDescription}${Preposition}${KeyDescription} as in "keywordForBrand" or "meterIDByMeter" or "queueTypeFromActionType", and sometimes omits the key part as in "ephemera" or "presences"_). No change needed here, but I'm musing if we should have such a helper.

Reply:
Not sure if it is the right thing, but please take a look

Reply:
I would say it's extremely close, and perfectly covers the narrow cases (cf. [`weakMemoizeUnary`](https://github.com/Agoric/agoric-sdk/blob/a253f74729e1001bf6e7302bb7da978a9f700a1e/packages/notifier/src/publish-kit.js#L70) and [`makeEphemeraProvider`](https://github.com/Agoric/agoric-sdk/blob/a253f74729e1001bf6e7302bb7da978a9f700a1e/packages/zoe/src/contractSupport/durability.js#L14), which are _identical_ modulo `harden()`ing of the returned function). The cases not covered are situations like here (where memoization is used in a "provide" sense and client code needs access to the map for overwriting entries) and things like CapData slot translation (where client code needs the ability to perform read-only peeking).